### PR TITLE
bugfix/issue-16-fvar-spec-quad_form_sym

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Summary:

Introduces fvar specialization of quad_form_sym

#### Intended Effect:

Admits higher-order autodiff of quad_form_sym for mixed type arguments (e.g. Matrix<double, -1, -1>, Matrix<fvar<var>, -1, -1>). Allows for fvar<T> instantiation of gaussian_dlm_obs_log.

#### How to Verify:

./runTests.py test/unit/math/mix/mat/fun/quad_form_test.cpp
./runTests.py test/unit/math/fwd/mat/fun/quad_form_test.cpp

These should fail in the current math branch.

#### Side Effects:

None.

#### Documentation:

None.

#### Reviewer Suggestions:

Anyone

### CODE REVIEW

File: stan/math/fwd/mat/fun/quad_form_sym.hpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include <stan/math/prim/mat/fun/quad_form_sym.hpp>
+
+namespace stan {
+  namespace math {
+
+    template<int RA, int CA, int RB, int CB, typename T>
+    inline Eigen::Matrix<stan::math::fvar<T>, CB, CB>
+    quad_form_sym(const Eigen::Matrix<stan::math::fvar<T>, RA, CA>& A,
+                  const Eigen::Matrix<double, RB, CB>& B) {
+      using stan::math::multiply;
```
Comment (syclik) [inline stan/math/fwd/mat/fun/quad_form_sym.hpp]: remove the `using` statement. Unnecessary now that everything is in the same namespace.

File: stan/math/fwd/mat/fun/quad_form_sym.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+      stan::math::check_square("quad_form_sym", "A", A);
```
Comment (syclik) [inline stan/math/fwd/mat/fun/quad_form_sym.hpp]: remove the namespace qualification (`stan::math`)

File: test/unit/math/fwd/mat/fun/quad_form_test.cpp [hunk 3]
```diff
@@ -118,6 +119,47 @@ TEST(AgradFwdMatrixQuadForm, quad_form_sym_mat_fd) {
...
+  -3, -3,
+  5,  2;
+  ad << 2.0,  3.0, 4.0,   5.0, 
+  3.0, 10.0, 2.0,   2.0,
+  4.0,  2.0, 7.0,   1.0,
+  5.0,  2.0, 1.0, 112.0;
+  
+  // fvar<double> - fvar<double>
+  matrix_fd resd = quad_form_sym(ad,bd);
+}
```
Comment (syclik) [inline test/unit/math/fwd/mat/fun/quad_form_test.cpp]: did you mean to add a test here?

File: test/unit/math/fwd/mat/fun/quad_form_test.cpp [hunk 4]
```diff
@@ -118,6 +119,47 @@ TEST(AgradFwdMatrixQuadForm, quad_form_sym_mat_fd) {
...
```
Comment (syclik) [inline test/unit/math/fwd/mat/fun/quad_form_test.cpp]: add an EXPECT_* here?

File: test/unit/math/fwd/mat/fun/quad_form_test.cpp [hunk 5]
```diff
@@ -203,6 +245,42 @@ TEST(AgradFwdMatrixQuadForm, quad_form_sym_vec_fd) {
...
+  
+  bd << 100, 0, -3, 5;
+  ad << 2.0,  3.0, 4.0,   5.0, 
+  3.0, 10.0, 2.0,   2.0,
+  4.0,  2.0, 7.0,   1.0,
+  5.0,  2.0, 1.0, 112.0;
+  
+  // fvar<double> - fvar<double> 
+  res = quad_form_sym(ad,bd);
+}
```
Comment (syclik) [inline test/unit/math/fwd/mat/fun/quad_form_test.cpp]: add test

File: test/unit/math/fwd/mat/fun/quad_form_test.cpp [hunk 6]
```diff
@@ -203,6 +245,42 @@ TEST(AgradFwdMatrixQuadForm, quad_form_sym_vec_fd) {
...
```
Comment (syclik) [inline test/unit/math/fwd/mat/fun/quad_form_test.cpp]: add test

File: test/unit/math/fwd/mat/fun/quad_form_test.cpp (hunk 3 above)
Comment (rtrangucci) [inline test/unit/math/fwd/mat/fun/quad_form_test.cpp]: these are all just examples that I can instantiate something that couldn't be instantiated; these tests won't compile in current develop. I'll add EXPECT_NO_THROW(check_not_nan(resd)) tests in all of these

**Comment by rtrangucci** [comment]
@syclik not sure what happened here, fwd/mat/fun/quad_form_test.cpp passes on my machine.

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by rtrangucci** [comment]
@bob-carpenter @syclik this is ready for code review, should be very straightforward. The main file to focus on is: stan/math/fwd/mat/fun/quad_form_sym.hpp

**Comment by syclik** [comment]
Cool.

**Pull request merged**